compiler/src/dotty/tools/dotc/staging/DirectTypeOf.scala

```diff
@@ -0,0 +1,27 @@
+package dotty.tools.dotc.staging
+
+import dotty.tools.dotc.ast.{tpd, untpd}
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.Types._
+
+object DirectTypeOf:
+  import tpd.*
+
+  /** Matches `quoted.Type.of[x.Underlying](quotes)` and extracts the TermRef to `x` */
+  def unapply(body: Tree)(using Context): Option[TermRef] = body.tpe match
+    case tp @ TypeRef(x: TermRef, _) if tp.symbol == defn.QuotedType_splice => Some(x)
+    case _ => None
+
+  object Healed:
+    /** Matches `quoted.Type.of[@SplicedType type T = x.Underlying; T](quotes)` and extracts the TermRef to `x` */
+    def unapply(body: Tree)(using Context): Option[TermRef] =
+      body match
+        case Block(List(tdef: TypeDef), tpt: TypeTree) =>
+          tpt.tpe match
+            case tpe: TypeRef if tpe.typeSymbol == tdef.symbol =>
+              tdef.rhs.tpe.hiBound match
+                case tp @ TypeRef(x: TermRef, _) if tp.symbol == defn.QuotedType_splice => Some(x)
+                case _ => None
```

Comment:
Can we do 
```suggestion
              DirectTypeOf.unapply(tdef.rhs.tpe.hiBound)
```
? they look the same.